Doc update for git proxy type (#1696)
wso2/docs-choreo-dev: en/docs/develop-components/develop-proxy/develop-a-git-api-proxy.md

         - Read access to issues and metadata.
        
        You can [revoke access](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/reviewing-your-authorized-integrations#reviewing-your-authorized-github-apps) if needed. Write access is only used for sending pull requests; Choreo will not push changes directly to your repository.    

6. Enter the following repository details:

    | **Field**              | **Value**          |
    |------------------------|--------------------|
    | **Organization**       | Your GitHub account|
    | **Repository**         | choreo-samples     |
    | **Branch**             | **`main`**         |
    | **API Directory**      | /choreo-samples/proxy-from-github/department-service |

7. Specify the following values as API proxy details:

    !!! info
        The **Context** field must be unique and cannot be changed after creation.**Version** and **Target** are mandatory fields. **Target** can be changed at any time after the creation.

    | **Field**       | **Value**                                  |
    |-----------------|--------------------------------------------|

>     | **Context**     | `department-service/v1`                                   |

Reviewer: Why are we setting the version again?
```suggestion
    | **Context**     | `/department-service`                                   |
```

>     | **Version**     | `V1.0`                                      |

Reviewer: ```suggestion
    | **Version**     | `v1.0`                                      |
```

> # Develop an API Proxy from Git Repository

Reviewer 2: Let's decide on https://github.com/wso2/docs-choreo-dev/pull/1696/files#r1976915437 and update this title and markdown file name accordingly if we are ok to go with the suggestion.

> # Develop an API Proxy from a GitHub Repository Source

Reviewer 2: Make sure to follow the .md file naming convention https://github.com/wso2/docs-choreo-dev/wiki/Markdown-File-Naming-Conventions-and-Content-Formatting#file-naming-convention
Need to rename this file:
`develop-a-git-api-proxy.md` to `develop-an-api-proxy-from-a-github-repository-source.md`

Let's rename the `develop-an-proxy.md` also accordingly and do the necessary file name changes in the mkdocs.yml

An API proxy acts as an intermediary between an existing API and Choreo, intercepting all requests made to the API. It also functions as a managed API, allowing you to apply essential API management features such as security policies and rate limiting.

In this guide, you will:

- Create an API proxy component to expose an existing API.
- Deploy the API proxy.
- Test the API proxy to verify its functionality.
- Manage the API.
- Consume the API.

## Prerequisites

- If you're signing in to the Choreo Console for the first time, create an organization:
    1. Go to [https://console.choreo.dev/](https://console.choreo.dev/) and sign in using your preferred method.
    2. Enter a unique organization name. For example, `Stark Industries`.
    3. Read and accept the privacy policy and terms of use.
    4. Click **Create**.

    This creates the organization and opens the **Project Home** page of the default project created for you.

> 

Reviewer 2: Since we are using a sample from the Choreo samples repo in this guide, we would need to ask users to  fork the repo and mention that the repo will have the Proxy from GitHub sample.

Author: what would the better way to do that ? an info banner asking to fork the sample repo ..

Author: No need to fork the repo , right. Anyone can use the repo without forking it. But when they need to change the exiting source , then only they need to fork it.

Author: Mention the step as a prerequisite

## Step 1: Create an API proxy

> To create an API proxy, you can either choose from a github repository, upload an OpenAPI specification or provide an OpenAPI specification URL. In this guide, you will choose the API definition from a Git source like Github repository.

Reviewer 2: Make sure to use the correct capitalization for "GitHub" in all occurrences.

Also, at the moment users can only select from GitHub right? Other Git repositories are not supported as of now isn't it?

Suggestion:
You can create an API proxy either by selecting the source from a GitHub repository, uploading an OpenAPI specification file, or providing an OpenAPI specification URL. This guide demonstrates how to create an API proxy using a GitHub repository as the source.

Author: `Also, at the moment users can only select from GitHub right? Other Git repositories are not supported as of now isn't it?`
Yes.

1. Go to [https://console.choreo.dev/](https://console.choreo.dev/) and sign in. This opens the project home page.
2. If you already have one or more components in your project, click **+ Create**. Otherwise, proceed to the next step.
3. Click the **API Proxy** card. This opens the **Create an API Proxy** pane.

> 4. You will see the three options to create a proxy component and here we select github option (selected by default).

Reviewer 2: This step seems out of place here. We need to mention this in step 6.

Author: Fixed

5. Specify the following values as component details:

    !!! info

>         The **Component Name** field must be unique and cannot be changed after creation.System will generate a unique name for you. But you are free to change the generated name.

Reviewer 2: ```suggestion
        The **Component Name** field must be unique and cannot be changed after creation. This value is automatically generated, but you can edit it if necessary.
```

        **Component Display Name** is a required field.

    | **Field**       | **Value**                                  |
    |-----------------|--------------------------------------------|
    | **Component Display Name**| `Department Service`                    |
    | **Component Name**        | `departmentService`                    |

>     | **Description**           | `This is a sample pet store`     |

Reviewer 2: Need to update the description to align with the sample used here. Pet store is incorrect.

6. Specify the following values as API proxy details:

    !!! info
        The **Context** field must be unique and cannot be changed after creation.**Version** and **Target** are mandatory fields. **Target** can be changed at any time after the creation.

    | **Field**       | **Value**                                  |
    |-----------------|--------------------------------------------|
    | **Context**     | `department-service`                                   |
    | **Version**     | `v1.0`                                      |
    | **Target**      | `https://samples.choreoapps.dev/company/hr`|

7. Specify the following values as component details:

    !!! info
        The **Component Name** field must be unique and cannot be changed after creation. This value is automatically generated, but you can edit it if necessary.
        **Component Display Name** is a required field.

    | **Field**       | **Value**                                  |
    |-----------------|--------------------------------------------|
    | **Component Display Name**| `Department Service`                    |

>     | **Component Name**        | `departmentService`                    |

Reviewer 2: Shall we go with `department- service`? There is a validation error when uppercase letters are used.

Reviewer: Lets use the automatically suggested name on the UI - "department-service"

> 6. Click **Create**. This creates the API proxy component and takes you to the **Overview** page.

Reviewer 2: This took me to the **Build** page.

Is it supposed to take me to the **Overview** page?

Author: This should be corrected to `Build`. Because git based proxies has a build option.

> 8. Click **Create**. This creates the API proxy component and takes you to the **Build** page.

Reviewer 2: I was trying out the steps and although component creation happened there was a glitch and I was not taken to the **Build** page. On deleting the component and trying again I get this validation error:
<img width="1537" alt="Screenshot 2025-03-11 at 07 57 43" src="https://github.com/user-attachments/assets/81d26168-24aa-49ad-b9fe-919631f26364" />
Let's test the steps end-to-end and get the doc updated accordingly. I might not have time to do this :)

Reviewer 3: the height of the Context input has to be fixed. @shanakama Height should be fixed always.

Author: Fix : https://github.com/wso2-enterprise/choreo-console/pull/10450

>     When you creating an API proxy by **pointing a git repository** , the Git source is considered as the **single source of truth**. So , the only way is to change API resources like resource addition, deletion of modifications should be done trought the Git source.

Reviewer 2: Suggestion:
When you create an API proxy from a GitHub repository source, the GitHub source serves as the single source of truth. Therefore, any modifications, such as adding or deleting resources, must be made through the GitHub repository.

## Step 2: Build

!!! info

>     There is a initial build running immediately once after you create a proxy.

Reviewer 2: Suggestion:
An initial build starts automatically as soon as you create the API proxy.

> 1. On the project home page, click on the `Department Service` component you created. This takes you to the component overview page.

Reviewer 2: If proxy component creation takes users to the **Build** page, we wouldn't need 1 and 2.
Better to try out the flow and update steps accordingly.

Author: Git based proxies should have a build page.

> 1. On the project home page, click on the `Department Service` component you created. This takes you to the component overview page.

Reviewer 2: Since the user is taken to the build page on creating the component, we don't need these steps right? We can say perhaps just mention that "The initial build starts automatically as soon as you create the API proxy. The build process may take some time. You can track progress in the **Build Details** pane. Once complete, the build status changes to **Success**."

Reviewer: @shanakama did we address this comment when resolving?

Author: I think we should keep this step. Even the user automatically redirect to the build page once after the creation completed , that is one time behavior. All other times , user should follow above steps to navigate to the build page.

!!! note
    The build process may take some time. You can track progress in the **Build Details** pane. Once complete, the build status changes to **Success**.

## Step 3: Deploy the API proxy

1. In the left navigation menu, click **Deploy**.
2. In the **Build Area** card, click **Configure & Deploy**. This opens the **Configure & Deploy** pane.

> 3. Select **External** as the API access mode and click **Deploy**. The **Development** card indicates the **Deployment Status** as **Active** when the API proxy is successfully deployed.

Reviewer 2: ```suggestion
3. Select **External** as the **API Access Mode** and click **Deploy**. The **Development** card indicates the **Deployment Status** as **Active** when the API proxy is successfully deployed.
```

            - To disable security for a specific resource, expand the relevant resource and turn off the **Security** toggle.
        4. Click **Apply**.

1. In the left navigation menu, click **Test** and then click **OpenAPI Console**.
2. Select **Development** from the environment drop-down list.
3. Expand the `GET /{departmentId}` resource and click **Try it Out**.
4. Enter `1` as the **departmentId** and click **Execute**. You will see a response similar to the following:

    ![API proxy response](../../assets/img/develop-components/develop-a-rest-api-proxy/rest-api-proxy-response.png){.cInlineImage-full}

    This indicates that your API proxy is working as expected.

## Step 5: Manage the API proxy

Now that you have a tested API proxy, you can publish it and make it available for application developers to consume. In this guide, you will apply rate limiting to the API and publish it.

### Step 5.1: Apply rate limiting to the API proxy

1. In the left navigation menu, click **Deploy**.
2. Go to the required environment card and click the settings icon corresponding to **API Configuration**.

> 3. In the **API Configuration** pane, click **Rate Limiting** to expand the section.

Reviewer: "API Configuration" is now renamed as "CORS, Rate Limiting and Resiliency"

> 4. Select **API Level** as the **Rate Limiting Level**.

Reviewer 2: Since there are UI changes, let's test the steps and update this section accordingly.

Author: Added changes here : https://github.com/wso2/docs-choreo-dev/pull/1696/commits/e071703f82e8f2a76a74ca7aac84ad280d696894

5. Specify appropriate values for the **Request Limit** and **Time Unit** fields. You can proceed with the default values.
6. Click **Apply**. This applies the rate limiting level to the API proxy and redeploys it.

### Step 5.2: Publish the API proxy

1. In the left navigation menu, click **Lifecycle** under **Manage**. This takes you to the **Lifecycle** page.
2. Click **Publish**.
3. In the **Publish API** dialog, click **Confirm** to proceed with publishing the API. If you want to change the display name, make the necessary changes and then click **Confirm**. This changes the API lifecycle state to **Published**.

## Step 6: Invoke the API

To generate credentials for the published API and invoke it via the Choreo Developer Portal, follow these steps:

> 1. In the **Lifecycle** page, click **Go to Devportal**. This takes you to the `HR API` in the Choreo Developer Portal.

Reviewer: This takes you to the `Department Service`?

2. **Generate Credentials**:
    1. In the Developer Portal left navigation menu, click **Production** under **Credentials**.
    2. Click **Generate Credentials**. Choreo generates new tokens and populates the **Consumer Key** and **Consumer Secret** fields.

    !!! tip
        To test the API via an API test tool or through code, click **Generate Access Token** and copy the test token. Alternatively, click **cURL** and copy the generated cURL command to use via a cURL client. You do not need to generate an access token if you are testing the API via the **Try Out** capability in the Choreo Developer Portal.

3. **Invoke the API**:
    1. In the Developer Portal left navigation menu, click **Try Out**.
    2. In the **Endpoint** list, select **Development** as the environment to try out the API.
    3. Click **Get Test Key** to generate an access token.
    4. Expand the `GET /{departmentId}` resource and click **Try it out**.
    5. Enter `1` as the **departmentId** and click **Execute**. You will see a response similar to the following:

        ![Try out response](../../assets/img/develop-components/develop-a-rest-api-proxy/try-out-response.png){.cInlineImage-full}

> Now, you have gained hands-on experience creating, deploying, testing, and publishing an API proxy using Choreo API Manager.

Reviewer: ```suggestion
Now, you have gained hands-on experience creating, deploying, testing, and publishing an API proxy using a Github repository.
```

Reviewer: Please fix this in the original too. It doesn't make sense to mention about Choreo API Manager as it looks like a product.

> To learn more about the API management capabilities supported by Choreo API Manager, see [API Management](../../api-management/lifecycle-management.md).

Reviewer: We are pointing to a wrong link it seems. We indicate about navigating to a generic feature page but we are directing to a lifecycle management page which is just a specific feature. If we don't have a better link to navigate, better to remove this line.

Reviewer: Need to fix this in the original too.
